library(ggplot2)
library(reshape2)
library(knitr)
```{r setup}
opts_chunk$set(fig.width=10, fig.height=9)
opts_chunk$set(dev=c('png'))


``` 

```{r}
reorder <- function(M,new_order) {
  M[lower.tri(M)] = t(M)[lower.tri(M)]
  M <- M[new_order,new_order]
  M[lower.tri(M)] <- NA
  M
}

sym <- function(M) {
  M[lower.tri(M)] = t(M)[lower.tri(M)]
  M
}


ggcolour <- function(n) {
  hues = seq(15, 375, length=n+1)
  hcl(h=hues, l=65, c=100)[1:n]
}


tree_order <- c(
"Th166.12", "Th245.13", "Th211.13" ,"Th246.13", "Th092.13" ,
"Th086.07", "Th106.11", "Th117.11", "Th134.11", "Th106.09", "Th074.13", "Th162.12", "Th132.11","Th230.12","Th196.12", 
"Th061.13", "Th095.13","Th068.12"
)
clades <- c(
  rep(2,5),
  rep(3,10),
  rep(1,3))
names(clades)<-tree_order
clades

```

```{r}

#INDEL_dist <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.INDEL.recode.vcf.dist.tab.txt",header=T,row.names=1,sep="\t")
#SNP_dist <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.SNP.recode.vcf.dist.tab.txt",header=T,row.names=1,sep="\t")

total_dist <- read.table("3D7DD2.REFCALL.noMinor.NoLaneErr.IncIncons.dist.tab.txt",header=T,row.names=1,sep="\t")
ratios_tab <- read.table("3D7DD2.REFCALL.noMinor.NoLaneErr.IncIncons.indel-snp.tab.txt",header=T,row.names=1,sep="\t")
tstv_tab <- read.table("3D7DD2.REFCALL.noMinor.NoLaneErr.IncIncons.tstv.tab.txt",header=T,row.names=1,sep="\t")


# INDEL_dist <- reorder(INDEL_dist,tree_order)
# SNP_dist <- reorder(SNP_dist,tree_order)
# total_dist <- reorder(total_dist,tree_order)



```



```{r}


# distances <- merge(
#   merge(
#     melt(as.matrix(INDEL_dist),value.name ="indel"),
#     melt(as.matrix(SNP_dist),value.name ="snp"),
#     by=c("Var1","Var2")
#     ),
#   melt(as.matrix(total_dist),value.name ="total"),
#   by=c("Var1","Var2")
# )
distances <-  melt(as.matrix(total_dist),value.name ="total")
ratios <-  melt(as.matrix(ratios_tab),value.name ="ratio")
tstv <-  melt(as.matrix(tstv_tab),value.name ="tstv")
ratios$Var2<-gsub('\\.','-',ratios$Var2)
distances$Var2 <- gsub('\\.','-',distances$Var2)
tstv$Var2 <- gsub('\\.','-',tstv$Var2)

distances$Var1 <- as.character(distances$Var1)
distances$Var2 <- as.character(distances$Var2)
ratios$Var1 <- as.character(ratios$Var1)
ratios$Var2 <- as.character(ratios$Var2)
tstv$Var1 <- as.character(tstv$Var1)
tstv$Var2 <- as.character(tstv$Var2)

#merge(distances,ratios,by=c("Var1","Var2"))
#distances <- distances[distances$Var1!=distances$Var2,]
distances <- merge(distances,ratios,by=c("Var1","Var2"))
distances <- merge(distances,tstv,by=c("Var1","Var2"))
distances <- distances[!is.na(distances$total),]

#distances$related <- distances$total < 10000
distances$related[distances$total >= 10000] <- "UNRELATED"
distances$related[distances$total < 10000] <- "RELATED"


#distances$Var1 <- factor(distances$Var1,levels=rev(tree_order))
#distances$Var2 <- factor(distances$Var2,levels=rev(tree_order))

```

```{r}
# distances$ratioP <- NA
# ID_unrelated <- c(sum(distances$indel[distances$related=="UNRELATED"]),sum(distances$snp[distances$related=="UNRELATED"]))
# ID_related <- c(sum(distances$indel[distances$related=="RELATED"]),sum(distances$snp[distances$related=="RELATED"]))
# tab <- cbind(ID_unrelated,ID_related)
# colnames(tab)<-c("unrelated","related"); rownames(tab) <- c("indel","snp")
# tab
# fisher.test(tab)
# 
# for(i in 1:dim(distances)[[1]]){ 
#   t <- fisher.test(t(rbind(ID_unrelated,distances[i,c("indel","snp")])))
#   distances[i,"ratioP"] <- t$p.value
#   }
# 
# distances$ratio[is.nan(distances$ratio)] <- NA
# #distances

```

```{r}

relcol <- scale_color_manual(values=c("white","black"))
vxlab <- theme(axis.text.x = element_text(angle = 90, hjust = 1))  

ggplot(distances,aes(x=Var1,y=Var2,fill=total,label=total,colour=related)) + geom_tile() + scale_fill_gradient(trans="log") + geom_text(size=3) + relcol + vxlab
ggplot(subset(distances,related=="RELATED"),aes(x=Var1,y=Var2,fill=ratio,label=ratio)) + geom_tile(size=1) + 
  geom_text(aes(colour="white"),size=5) + 
  geom_text(data=subset(distances,related=="UNRELATED"),size=4) + sigcol + vxlab + ggtitle("SNP:INDEL")
ggplot(subset(distances,related=="RELATED"),aes(x=Var1,y=Var2,fill=tstv,label=tstv)) + geom_tile(size=1) + 
  geom_text(aes(colour="white"),size=5) + 
  geom_text(data=subset(distances,related=="UNRELATED"),size=4) + sigcol + vxlab + ggtitle("ts:tv")

# ggplot(subset(distances,related=="RELATED"),aes(x=Var1,y=Var2,fill=total,label=ratio_t)) + geom_tile(size=1) + 
#   geom_text(aes(colour=ratioP < 0.01),size=4) + 
#   sigcol + vxlab
ggplot(subset(distances,related=="RELATED"),aes(x=Var1,y=Var2,fill=interval_yrs,label=ratio_t)) + geom_tile(size=1) + 
  geom_text(aes(colour=ratioP < 0.01),size=4) + 
  sigcol + vxlab

```

```{r}
hets <- read.table("Thies_all_manual_2.PASS.Cls.miss0.5.LMRG.het.HETCOUNTS",header=T)
colnames(hets)[6] <- "hets"
rownames(hets) <- hets$INDV
hets <- hets[tree_order,]
ggplot(hets,aes(x=INDV,y=hets)) + geom_bar(stat="identity") + coord_flip()

sings <- read.table("Thies_all_manual_2.PASS.Cls.miss0.5.LMRG.singletons",header=T)
sings <- subset(sings,SINGLETON.DOUBLETON=="D")
#sings$vartype<-"SNP"
#sings$vartype[apply(sings$ALLELE,1,FUN=length)>1]<-"INDEL"
#table(sings[,c("INDV","vartype")])
sings<-table(sings[,c("INDV")])
sings <- sings[tree_order]
hets$sings <- sings 
hets$clade <- as.factor(clades)

hets$INDV <- factor(hets$INDV,levels = rev(tree_order))
cftab <- melt(hets[,c("INDV","clade","hets","sings")],id.vars=c("INDV","clade"))

ggplot(cftab,aes(x=INDV,y=value,fill=clade)) + geom_bar(stat="identity") + coord_flip() + facet_grid(. ~ variable,scales ="free_y") + scale_fill_manual(values=c("red","blue","green"))


```


```{r}
distances$ratPSig <- distances$ratioP<0.01
distcfs <- melt(distances[,c("snp","indel","total","ratPSig")],id.vars=c("total","ratPSig"))
ggplot(distcfs,aes(x=total,y=value,colour=variable,group=variable)) + geom_point() + scale_y_log10() + scale_x_log10()
#ggplot(distcfs,aes(x=total,y=value,colour=variable,group=variable)) + geom_point() + geom_smooth(method = "lm", se = FALSE)+ xlim(0,1000)+ylim(0,1000)

maxv = 750
ggplot(distcfs,aes(x=total,y=value,colour=variable,group=variable,shape=ratPSig)) + geom_point() + 
  geom_smooth(se = FALSE) + 
  geom_smooth(method="lm",se = FALSE,colour="grey")+  
  xlim(0,maxv)+ylim(0,maxv)
maxv = 400
ggplot(distcfs,aes(x=total,y=value,colour=variable,group=variable,shape=ratPSig)) + geom_point() + 
  geom_smooth(se = FALSE) + 
  geom_smooth(method="lm",se = FALSE,colour="grey")+  
  xlim(0,maxv)+ylim(0,maxv)
```

