```{r setup}
library(ggplot2)
library(hexbin)
library(reshape2)
library(knitr)

library(gridExtra)
library(grid)
library(ggplot2)
library(lattice)
opts_chunk$set(fig.width=20, fig.height=8,dev=c('png','postscript'),warning=F)
#opts_chunk$set(fig.width=20, fig.height=8,dev=c('png'),warning=F)

``` 

```{r}
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}

reflect<-function(mat){
  mat[lower.tri(mat)] <- t(mat)[lower.tri(mat)]
  return(mat)
  }

tree_order <- c(
  "Th162.12", "Th230.12","Th196.12","Th132.11", "Th074.13", "Th106.09", "Th117.11", "Th134.11", "Th106.11","Th086.07",   
  "Th211.13","Th246.13", "Th092.13","Th166.12", "Th245.13", 
  "Th061.13", "Th095.13", "Th068.12"
)
clades <- c(
  rep(3,10),
  rep(2,5),
  rep(1,3))
names(clades)<-tree_order
clades


```


```{r}
callboth <- read.table("thies_disco.callAll.FILT.m0.5.dist.tab.txt",sep="\t")
callboth <- reflect(callboth)
callboth$from=rownames(callboth)
callboth <- melt(callboth,variable.name = "to",value.name="both")
callMax <- read.table("thies_disco.FILT.dist.tab.txt",sep="\t")
callMax <- reflect(callMax)
callMax$from=rownames(callMax)
callMax <- melt(callMax,variable.name = "to",value.name="max")


callsnp <- read.table("thies_disco.FILT.m0.5.SNP.tab.txt",sep="\t")
callsnp$from=rownames(callsnp )
callsnp  <- melt(callsnp ,variable.name = "to",value.name="snp")

callindel <- read.table("thies_disco.FILT.m0.5.INDEL.tab.txt",sep="\t")
callindel $from=rownames(callindel )
callindel  <- melt(callindel ,variable.name = "to",value.name="indel")


callCfDisco <- merge(merge(callboth,callMax),merge(callsnp,callindel))
callCfDisco <- callCfDisco[!is.na(callCfDisco$snp),]
callCfDisco <- callCfDisco[callCfDisco$from != callCfDisco$to,]

#callCfDisco <- callCfDisco[callCfDisco$snp <= 1000,]
callCfDisco$prog <- "disco"
```


```{r}

callbothGATK100 <- read.table("thies_haplo100.callAll.FILT.m0.5.dist.tab.txt",sep="\t")
callbothGATK100 <- reflect(callbothGATK100)
callbothGATK100$from=rownames(callbothGATK100)
callbothGATK100 <- melt(callbothGATK100,variable.name = "to",value.name="both")
callMaxGATK100 <- read.table("thies_haplo100.FILT.m0.5.dist.tab.txt",sep="\t")
callMaxGATK100 <- reflect(callMaxGATK100)
callMaxGATK100$from=rownames(callMaxGATK100)
callMaxGATK100 <- melt(callMaxGATK100,variable.name = "to",value.name="max")

callsnpGATK100 <- read.table("thies_haplo100.FILT.m0.5.SNP.tab.txt",sep="\t")
callsnpGATK100$from=rownames(callsnpGATK100)
callsnpGATK100 <- melt(callsnpGATK100,variable.name = "to",value.name="snp")
callindelGATK100 <- read.table("thies_haplo100.FILT.m0.5.INDEL.tab.txt",sep="\t")
callindelGATK100$from=rownames(callindelGATK100)
callindelGATK100 <- melt(callindelGATK100,variable.name = "to",value.name="indel")


callCfGATK100 <- merge(merge(callsnpGATK100,callindelGATK100),merge(callbothGATK100,callMaxGATK100))
callCfGATK100 <- callCfGATK100[!is.na(callCfGATK100$both),]
callCfGATK100 <- callCfGATK100[callCfGATK100$from != callCfGATK100$to,]
head(callCfGATK100)

#callCfGATK100 <- callCfGATK100[callCfGATK100$both > 0,]

#callCfGATK100 <- callCfGATK100[callCfGATK100$both <= 5000,]
callCfGATK100$prog <- "haplo100"

```


```{r}

callbothGATK250 <- read.table("thies_haplo250.callAll.FILT.m0.5.dist.tab.txt",sep="\t")
callbothGATK250 <- reflect(callbothGATK250)
callbothGATK250$from=rownames(callbothGATK250)
callbothGATK250 <- melt(callbothGATK250,variable.name = "to",value.name="both")
callMaxGATK250 <- read.table("thies_haplo250.FILT.m0.5.dist.tab.txt",sep="\t")
callMaxGATK250 <- reflect(callMaxGATK250)
callMaxGATK250$from=rownames(callMaxGATK250)
callMaxGATK250 <- melt(callMaxGATK250,variable.name = "to",value.name="max")

callsnpGATK250 <- read.table("thies_haplo250.FILT.m0.5.SNP.tab.txt",sep="\t")
callsnpGATK250$from=rownames(callsnpGATK250)
callsnpGATK250 <- melt(callsnpGATK250,variable.name = "to",value.name="snp")
callindelGATK250 <- read.table("thies_haplo250.FILT.m0.5.INDEL.tab.txt",sep="\t")
callindelGATK250$from=rownames(callindelGATK250)
callindelGATK250 <- melt(callindelGATK250,variable.name = "to",value.name="indel")


callCfGATK250 <- merge(merge(callsnpGATK250,callindelGATK250),merge(callbothGATK250,callMaxGATK250))
callCfGATK250 <- callCfGATK250[!is.na(callCfGATK250$both),]
callCfGATK250 <- callCfGATK250[callCfGATK250$from != callCfGATK250$to,]
head(callCfGATK250)

#callCfGATK250 <- callCfGATK250[callCfGATK250$both > 0,]

#callCfGATK250 <- callCfGATK250[callCfGATK250$both <= 5000,]
callCfGATK250$prog <- "haplo250"

```




```{r}

callCf <- rbind(callCfDisco,callCfGATK100,callCfGATK250)


samples <- read.table("Thies.samples.txt",col.names=c("ID","lane","dataset","sample","bam"))
samples <- unique(samples[,c("ID","sample")])
snames <- samples[,c("ID")]; samples<- as.character(samples[,c("sample")]) ; names(samples) <- snames
callCf$from[! callCf$from %in% samples] <- samples[as.character(callCf$from[! callCf$from %in% samples])]
names(samples) <- gsub('-',".",snames)
callCf$to[! callCf$to %in% samples] <- samples[as.character(callCf$to[! callCf$to %in% samples])]

callCf$to <- factor(callCf$to,levels=tree_order,ordered=T)
callCf$from <- factor(callCf$from,levels=tree_order,ordered=T)
callCfDisco$to <- factor(callCfDisco$to,levels=tree_order,ordered=T)
callCfDisco$from <- factor(callCfDisco$from,levels=tree_order,ordered=T)

callCf$related=clades[callCf$from]==clades[callCf$to]
callCf$extended <- callCf$max-callCf$both

callCf$prog <- factor(callCf$prog,levels=c("disco","haplo250","haplo100"),ordered=T)
```

```{r}

callCf <- subset(callCf,related==T)
discoSI <- cor.test(callCf[callCf$prog=="disco","snp"],callCf[callCf$prog=="disco","indel"])
haplo100SI <- cor.test(callCf[callCf$prog=="haplo100","snp"],callCf[callCf$prog=="haplo100","indel"])
haplo250SI <- cor.test(callCf[callCf$prog=="haplo250","snp"],callCf[callCf$prog=="haplo250","indel"])
discoCM <- cor.test(callCf[callCf$prog=="disco","both"],callCf[callCf$prog=="disco","max"])
haplo100CM <- cor.test(callCf[callCf$prog=="haplo100","both"],callCf[callCf$prog=="haplo100","max"])
haplo250CM <- cor.test(callCf[callCf$prog=="haplo250","both"],callCf[callCf$prog=="haplo250","max"])
discoCE <- cor.test(callCf[callCf$prog=="disco","both"],callCf[callCf$prog=="disco","extended"])
haplo100CE <- cor.test(callCf[callCf$prog=="haplo100","both"],callCf[callCf$prog=="haplo100","extended"])
haplo250CE <- cor.test(callCf[callCf$prog=="haplo250","both"],callCf[callCf$prog=="haplo250","extended"])

corrs <- data.frame("prog"=c("disco","haplo100","haplo250"),
          "SI" = c(discoSI$estimate,haplo100SI$estimate,haplo250SI$estimate),
          "CM" = c(discoCM$estimate,haplo100CM$estimate,haplo250CM$estimate),
          "CE" = c(discoCE$estimate,haplo100CE$estimate,haplo250CE$estimate))


discoSI
ggplot(subset(callCf,prog=="disco"),aes(y=snp,x=indel)) + geom_point() +  
  geom_label(y=100,x=150,label=paste("r = ",round(discoSI$estimate,2)),size=5) +
  ggtitle("Discovar calls - SNP:INDEL correlation")

discoCM
ggplot(subset(callCf,prog=="disco"),aes(y=both,x=max)) + geom_point() +  
  geom_label(y=100,x=150,label=paste("r = ",round(discoCM$estimate,2)),size=5) +
  ggtitle("Discovar calls - accessible:inaccessible correlation")
#ggplot(callCf,aes(x=both,y=haplo)) + geom_point(aes(y=disco),colour="red")



ggplot(callCf,aes(y=snp,x=indel)) + geom_point() + geom_smooth(method = lm,se = F,linetype=2) +
  facet_grid(. ~ prog)+
  ggtitle("SNP:INDEL correlation")+
  geom_label(data=corrs,aes(y=100,x=750,label=paste("r = ",round(SI,2))),size=5) +
  theme(legend.position="bottom") + xlab("extended genome") + ylab("core genome")

# ggplot(callCf,aes(y=both,x=max)) + geom_point() + geom_smooth(method = lm,se = F,linetype=2) +
#   facet_grid(. ~ prog)+
#   ggtitle("CORE:MAX correlation")+
#   geom_label(data=corrs,aes(y=100,x=1000,label=paste("r = ",round(CM,3))),size=5) +
#   theme(legend.position="bottom") + xlab("extended genome") + ylab("core genome")

ggplot(callCf,aes(y=both,x=extended)) + geom_point() + geom_smooth(method = lm,se = F,linetype=2) +
  facet_grid(. ~ prog)+
  ggtitle("CORE:EXTENDED correlation")+
  geom_label(data=corrs,aes(y=100,x=250,label=paste("r = ",round(CE,2))),size=5) +
  theme(legend.position="bottom") + xlab("extended genome") + ylab("core genome")


```


```{r}
levels(callCf$to)
levels(callCf$from)
length(callCf$to>callCf$from)
sum(callCf$to>callCf$from)
cbind(callCf$to,callCf$from,callCf$to>=callCf$from)
callCf <- callCf[callCf$to >= callCf$from,]


dcol = scale_fill_gradient(low="blue",high="black")
vxlab <- theme(axis.text.x = element_text(angle = -90, hjust = 1))  
vxleg <- theme(legend.position="bottom",legend.text = element_text(angle = -90,hjust=-1), axis.title=element_blank())
# ggplot(callCf) + 
#   geom_tile(aes(x=from,y=to,fill=extended),size=1) + 
#   geom_text(aes(x=from,y=to,label=extended), size=3,colour="white") + 
#   facet_grid(. ~ prog) + dcol + vxlab
# 
# ggplot(callCf) + 
#   geom_tile(aes(x=to,y=from,fill=both),size=1) + 
#   geom_text(aes(x=to,y=from,label=extended), size=3,colour="white") + 
#   facet_grid(. ~ prog) + dcol + vxlab
# 

callCfCX <- melt(callCf[,c("from","to","prog","related","both","extended")])
callCfCX <- subset(callCfCX,related==T)
callCfCX$from <- factor(callCfCX$from,levels=tree_order,ordered=T)
callCfCX$to <- factor(callCfCX$to,levels=tree_order,ordered=T)

callCfSI <- melt(callCf[,c("from","to","prog","related","snp","indel")])
callCfSI <- subset(callCfSI,related==T)
callCfSI$from <- factor(callCfSI$from,levels=tree_order,ordered=T)
callCfSI$to <- factor(callCfSI$to,levels=tree_order,ordered=T)


colscale <- (c(-500,-0.001,0,0.001,500)+500)/(500+500)
cxcol <- scale_fill_gradientn(colors = c("black","red","grey","dark green","black"),
                              values =colscale,na.value = "black",
                              breaks =c(-1000,-500,0,500,1000),
                              labels=c(500,250,0,500,1000),
                              limits=c(-1000,1000))

ggplot(data=callCfCX,aes(x=from,y=to,fill=value,label=value)) + 
  geom_tile(data=subset(callCfCX,variable=="both"),size=1) + 
  geom_text(data=subset(callCfCX,variable=="both"), size=3,colour="white") + 
  geom_tile(data=subset(callCfCX,variable=="extended"),aes(x=to,y=from,fill=value*-2),size=1) + 
  geom_text(data=subset(callCfCX,variable=="extended"),aes(x=to,y=from,label=value), size=3,colour="white") + 
  facet_grid(. ~ prog) + cxcol + vxlab + vxleg

iscol <- scale_fill_gradientn(colors = c("black","orange","grey","blue","black"),
                              values =colscale,na.value = "black",
                              breaks =c(-1000,-250,0,250,1000),
                              labels=c(1000,250,0,250,1000),
                              limits=c(-1000,1000))
ggplot(data=callCfSI,aes(x=from,y=to,fill=value,label=value)) + 
  geom_tile(data=subset(callCfSI,variable=="snp"),size=1) + 
  geom_text(data=subset(callCfSI,variable=="snp"), size=3,colour="white") + 
  geom_tile(data=subset(callCfSI,variable=="indel"),aes(x=to,y=from,fill=value*-1),size=1) + 
  geom_text(data=subset(callCfSI,variable=="indel"),aes(x=to,y=from,label=value), size=3,colour="white") + 
  facet_grid(. ~ prog) + iscol + vxlab + vxleg

```

