#seqtrack tutorial  
```{r}

library("phyloTop")
library("phylobase")
library("ape")
library("phytools")
library(knitr)
library(igraph)
library("RColorBrewer")
library(ggplot2)

opts_chunk$set(fig.width=9, fig.height=9)
opts_chunk$set(dev=c('png','postscript'))



```


```{r}
#thies <- read.nexus("sum_ALL.target.2.nexus")
tree <- read.newick("sum_ALL.target.newick")



tree$tip.label<- gsub("'","",tree$tip.label)


is.rooted(tree)

red <- c("Th061.13", "Th095.13", "Th068.12")
blue <- c("Th166.12", "Th245.13", "Th211.13", "Th246.13", "Th092.13")
green <- c("Th230.12","Th196.12","Th106.09","Th074.13","Th106.11","Th117.11","Th134.11","Th086.07","Th162.12","Th132.11")
  
  
tree
splitTop(tree,1)
splitTop(tree,2)

redtree <- extract.clade(tree,node=34)
bluetree <- extract.clade(tree,node=21)
greentree <- extract.clade(tree,node=25)


# subtree <- greentree
# sackin.phylo(subtree)
# widths(subtree)
# avgLadder(subtree)
# ILnumber(subtree)
# colless.phylo(subtree)
# ladderSizes(subtree)
# ladderShow(subtree)

phyloTop(list(bluetree,greentree,redtree))






```


```{r}
cophyloplot(Itree,Stree,cbind(Stree$tip.label,Stree$tip.label))

Stree <- read.nexus("Thies_SNP.green.nexus")
Itree <- read.nexus("Thies_INDEL.green.nexus")
dist.topo(Stree,Itree)

Imat <- as.dist(t(read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.INDEL.recode.vcf.dist.tab.txt",sep="\t")))
Smat <- as.dist(t(read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.SNP.recode.vcf.dist.tab.txt",sep="\t")))

dists <- data.frame("SNP"=as.numeric(Smat),"INDEL"=as.numeric(Imat))
dists <- dists[dists$SNP < 5000,]
# R^2 = 
cor(dists$SNP,dists$INDEL)^2
ggplot(dists,aes(x=INDEL,y=SNP)) + geom_point() + geom_smooth(method = glm,se = F)



```
