library(ggplot2)
library(hexbin)
library(reshape2)
library(knitr)

library(gridExtra)
library(grid)
library(ggplot2)
library(lattice)
```{r setup}
opts_chunk$set(fig.width=12, fig.height=8,dev='png',warning=F)

``` 

```{r}

discoFilt <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.Dd2-2D4_FILT_EVAL/weighted_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names=c("score","true_positives","false_positives","false_negatives","precision","sensitivity","f_measure"))
discoFilt$caller="disco"
discoFilt$region="GENOME"

discoFiltSNPs <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.Dd2-2D4_FILT_EVAL/snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
discoFiltINDELs <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.Dd2-2D4_FILT_EVAL/non_snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
discoFiltSNPs$type="SNP"
discoFiltINDELs$type="INDEL"
discoFiltType <- rbind(discoFiltSNPs,discoFiltINDELs)
discoFiltType$caller="disco"
discoFiltType$region="GENOME"
posns <- intersect(discoFiltSNPs$score,discoFiltINDELs$score)
discoFiltType <- subset(discoFiltType,score %in% posns)


discoCALLFilt <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.CALLBOTH.Dd2-2D4_FILT_EVAL/weighted_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names=c("score","true_positives","false_positives","false_negatives","precision","sensitivity","f_measure"))
discoCALLFilt$caller="disco"
discoCALLFilt$region="CALLABLE"

discoCALLFiltSNPs <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.CALLBOTH.Dd2-2D4_FILT_EVAL/snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
discoCALLFiltINDELs <- read.table("3D7DD2.REFCALL.PASS.DD2CONC.PFX.LMRG.HAP.CALLBOTH.Dd2-2D4_FILT_EVAL/non_snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
discoCALLFiltSNPs$type="SNP"
discoCALLFiltINDELs$type="INDEL"
discoCALLFiltType <- rbind(discoCALLFiltSNPs,discoCALLFiltINDELs)
discoCALLFiltType$caller="disco"
discoCALLFiltType$region="CALLABLE"
posns <- intersect(discoCALLFiltSNPs$score,discoCALLFiltINDELs$score)
discoCALLFiltType <- subset(discoCALLFiltType,score %in% posns)


head(discoFilt); head(discoCALLFilt)
head(discoFiltType); head(discoCALLFiltType)


# discoFilt <- read.table("disco_3D7_0901_weighted_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t")
# colnames(discoFilt) <-c("score","true_positives","false_positives","false_negatives","precision","sensitivity","f_measure")
# discoFilt$sample="0901"
# discoFilt$caller="disco"
# discoFilt$filter=T
```

```{r}
gatkFilt <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.Dd2-2D4_FILT_EVAL/weighted_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names=c("score","true_positives","false_positives","false_negatives","precision","sensitivity","f_measure"))
gatkFilt$caller="gatk"
gatkFilt$region="GENOME"

gatkFiltSNPs <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.Dd2-2D4_FILT_EVAL/snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
gatkFiltINDELs <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.Dd2-2D4_FILT_EVAL/non_snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
gatkFiltSNPs$type="SNP"
gatkFiltINDELs$type="INDEL"
gatkFiltType <- rbind(gatkFiltSNPs,gatkFiltINDELs)
gatkFiltType$caller="gatk"
gatkFiltType$region="GENOME"
posns <- intersect(gatkFiltSNPs$score,gatkFiltINDELs$score)
gatkFiltType <- subset(gatkFiltType,score %in% posns)


gatkCALLFilt <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.CALLBOTH.Dd2-2D4_FILT_EVAL/weighted_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names=c("score","true_positives","false_positives","false_negatives","precision","sensitivity","f_measure"))
gatkCALLFilt$caller="gatk"
gatkCALLFilt$region="CALLABLE"

gatkCALLFiltSNPs <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.CALLBOTH.Dd2-2D4_FILT_EVAL/snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
gatkCALLFiltINDELs <- read.table("3D7DD2_300100.VQSR.Pass90.RENAME.CALLBOTH.Dd2-2D4_FILT_EVAL/non_snp_roc.tsv.gz",header=F,skip = 3,stringsAsFactors =F,sep="\t",col.names = c("score","true_positives","false_positives"))
gatkCALLFiltSNPs$type="SNP"
gatkCALLFiltINDELs$type="INDEL"
gatkCALLFiltType <- rbind(gatkCALLFiltSNPs,gatkCALLFiltINDELs)
gatkCALLFiltType$caller="gatk"
gatkCALLFiltType$region="CALLABLE"
posns <- intersect(gatkCALLFiltSNPs$score,gatkCALLFiltINDELs$score)
gatkCALLFiltType <- subset(gatkCALLFiltType,score %in% posns)
```

```{r}

rocCf <- rbind(discoCALLFiltType,discoFiltType,gatkCALLFiltType,gatkFiltType)

```

```{r}
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}
```


```{r}

#ROC PLOT
ggplot(discoCALLFiltType,aes(x=false_positives,y=true_positives,colour=type)) + 
  geom_line() + facet_grid(. ~ caller) + theme(legend.position="bottom")

ggplot(discoFiltType,aes(x=false_positives,y=true_positives,colour=type)) + 
  geom_point() + facet_grid(. ~ caller) + theme(legend.position="bottom")


ggplot(discoFiltType,aes(x=false_positives,y=true_positives,colour=type)) + 
  geom_line() + facet_grid(. ~ caller) + theme(legend.position="bottom")


ggplot(rocCf,aes(x=false_positives,y=true_positives,colour=type,linetype=region)) + 
  geom_line() + facet_grid(. ~ caller) + theme(legend.position="bottom")

ggplot(rocCf,aes(x=false_positives,y=true_positives,colour=type,linetype=region)) + 
  geom_line() + scale_x_log10() + facet_grid(. ~ caller) + theme(legend.position="bottom")

```


```{r}
callboth <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.CALLBOTH.vcf.dist.tab.txt",sep="\t")
callboth$from=rownames(callboth)
callboth <- melt(callboth,variable.name = "to",value.name="both")
callDisco <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.CALLDISCOONLY.vcf.dist.tab.txt",sep="\t")
callDisco$from=rownames(callDisco)
callDisco <- melt(callDisco,variable.name = "to",value.name="disco")


callsnp <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.SNP.recode.vcf.dist.tab.txt",sep="\t")
callsnp $from=rownames(callsnp )
callsnp  <- melt(callsnp ,variable.name = "to",value.name="snp")

callindel <- read.table("Thies_all_manual.PASS.Cls.miss0.5.LMRG.HAP.INDEL.recode.vcf.dist.tab.txt",sep="\t")
callindel $from=rownames(callindel )
callindel  <- melt(callindel ,variable.name = "to",value.name="indel")


callCf <- merge(merge(callboth,callDisco),merge(callsnp,callindel))
callCf <- callCf[!is.na(callCf$snp),]
callCf <- callCf[callCf$from != callCf$to,]

callCf <- callCf[callCf$snp <= 1000,]

callbothGATK <- read.table("thies_300100_haplo.CALLBOTH.RENAME.dist.tab.txt",sep="\t")
callbothGATK$from=rownames(callbothGATK)
callbothGATK <- melt(callbothGATK,variable.name = "to",value.name="both")
callhaploGATK <- read.table("thies_300100_haplo.CALLHAPLO.RENAME.dist.tab.txt",sep="\t")
callhaploGATK$from=rownames(callhaploGATK)
callhaploGATK <- melt(callhaploGATK,variable.name = "to",value.name="haplo")
calldiscoGATK <- read.table("thies_300100_haplo.CALLDISCO.RENAME.dist.tab.txt",sep="\t")
calldiscoGATK$from=rownames(calldiscoGATK)
calldiscoGATK <- melt(calldiscoGATK,variable.name = "to",value.name="disco")

callsnpGATK <- read.table("thies_300100_haplo.CALLHAPLO.RENAME.SNP.dist.tab.txt",sep="\t")
callsnpGATK$from=rownames(callsnpGATK)
callsnpGATK <- melt(callsnpGATK,variable.name = "to",value.name="snp")
callindelGATK <- read.table("thies_300100_haplo.CALLHAPLO.RENAME.INDEL.dist.tab.txt",sep="\t")
callindelGATK$from=rownames(callindelGATK)
callindelGATK <- melt(callindelGATK,variable.name = "to",value.name="indel")


callCfGATK <- merge(merge(callsnpGATK,callindelGATK),merge(callbothGATK,callhaploGATK))
callCfGATK <- merge(callCfGATK,calldiscoGATK)
callCfGATK <- callCfGATK[!is.na(callCfGATK$both),]
callCfGATK <- callCfGATK[callCfGATK$from != callCfGATK$to,]
head(callCfGATK)

callCfGATK <- callCfGATK[callCfGATK$both > 0,]

callCfGATK <- callCfGATK[callCfGATK$both <= 5000,]

#remove core calls from disco/haplo regions
# discovar already called on disco only
#callCf$disco <- callCf$disco-callCf$both

callCfGATK$disco <- callCfGATK$disco-callCfGATK$both
callCfGATK$haplo <- callCfGATK$haplo-callCfGATK$both
```

```{r}


cor(callCf$snp,callCf$indel)
ggplot(callCf,aes(y=snp,x=indel)) + geom_point() +  
  geom_label(y=100,x=150,label=paste("r = ",round(cor(callCf$snp,callCf$indel),2)),size=5) +
  ggtitle("Discovar calls - SNP:INDEL correlation")

cor(callCf$both,callCf$disco)
ggplot(callCf,aes(y=both,x=disco)) + geom_point() +  
  geom_label(y=100,x=150,label=paste("r = ",round(cor(callCf$both,callCf$disco),2)),size=5) +
  ggtitle("Discovar calls - accessible:inaccessible correlation")
#ggplot(callCf,aes(x=both,y=haplo)) + geom_point(aes(y=disco),colour="red")

cor(callCfGATK$both,callCfGATK$disco)
cor(callCfGATK$both,callCfGATK$haplo)
ggplot(callCfGATK,aes(y=both,x=haplo)) + 
  geom_point(aes(x=haplo),colour="blue") + geom_point(aes(x=disco),colour="red") +
  geom_label(y=1500,x=375,label=paste("r = ",round(cor(callCfGATK$both,callCfGATK$haplo),2)),size=5,colour="blue") + geom_label(y=1000,x=750,label=paste("r = ",round(cor(callCfGATK$both,callCfGATK$disco),2)),size=5,colour="red") +
  ggtitle("GATK calls - accessible:inaccessible correlation")

cor(callCfGATK$snp,callCfGATK$indel)
ggplot(callCfGATK,aes(y=snp,x=indel)) + geom_point() + 
  geom_label(x=1500,y=300,label=paste("r = ",round(cor(callCfGATK$snp,callCfGATK$indel),2)),size=5) +
  ggtitle("GATK calls - SNP:INDEL correlation")


```


