library(ggplot2)
library(reshape2)
library(knitr)
```{r setup}
opts_chunk$set(fig.width=13, fig.height=12)

``` 

```{r}
getGenoTab <- function(filename) {
  genotab <- read.table(filename,sep="\t",header=T,stringsAsFactors=F,na.strings=c('.','. '))
  genotab <- unique(genotab)
  
  #remove ref calls
  #genotab <- genotab[rowSums(genotab[,5:dim(genotab)[2]],na.rm=T)!=0,]
  
  #colnames(genotab)[1:2]<-c("chr","pos")
  genotab.m <- melt(genotab,id.vars =c("chr","pos","type","alleles"),variable.name ="indv")
  genotab.m$value <- as.factor(genotab.m$value)
  #sort levels
  genotab.m$indv <- factor(genotab.m$indv,levels=sort(levels(genotab.m$indv)))
  
  genotab.m <- genotab.m[!is.na(genotab.m$value),]
  genotab.m$value <- as.numeric(as.character(genotab.m$value))
  
  hNo=0
  vlast=0
  genotab.m$haplo=0
  for (i in c(1:dim(genotab.m)[[1]])) {
    if (genotab.m[i,"value"]==0) {haplo=0}
    else if (genotab.m[i,"value"]>=1) {
      if (vlast==0) {hNo <- hNo+1}
        genotab.m[i,"haplo"]<-hNo  
      }
    vlast=genotab.m[i,"value"]
  }
  genotab.m$value <- factor(genotab.m$value)
  genotab.m$haplo <- factor(genotab.m$haplo)
  genotab.m
}

```

```{r}
clade1F <- "HFGR_all_incDD2select.PASS.miss0.5.LMRG.HAP.DISCORDS.MKSNGL.alleles.tab.txt"
cladeH2 <- "HFGR_all_incDD2select.PASS.miss0.5.LMRG.HAP.DISCORDS.MKSNGL.HFII.alleles.tab.txt"
cladeH3 <- "HFGR_all_incDD2select.PASS.miss0.5.LMRG.HAP.DISCORDS.MKSNGL.HFIII.alleles.tab.txt"

genotab1 <- getGenoTab(clade1F)
genotabH2 <- getGenoTab(cladeH2)
genotabH3 <- getGenoTab(cladeH3)

newOrder <- c("DD2.2D4", 
              "HFGRII.2.10x","HFGRII.1.30x"  , "HFGRII.1.200x" , "HFGRII.2.200x" , "HFGRII.2.300x"  ,
"HFGRIII.1.30x",  "HFGRIII.1.60x",  "HFGRIII.2.60x",  "HFGRIII.3.200x")
newOrder <- rev(newOrder)
genotab1$indv <- factor(genotab1$indv,levels = newOrder,ordered=T)
genotabH2$indv <- factor(genotabH2$indv,levels = newOrder,ordered=T)
genotabH3$indv <- factor(genotabH3$indv,levels = newOrder,ordered=T)
genotabH2 <- subset(genotabH2, indv %in% c("DD2.2D4", "HFGRII.2.10x","HFGRII.1.30x"  , "HFGRII.1.200x" , "HFGRII.2.200x" , "HFGRII.2.300x"))
genotabH3 <- subset(genotabH3, indv %in% c("DD2.2D4", "HFGRIII.1.30x",  "HFGRIII.1.60x",  "HFGRIII.2.60x",  "HFGRIII.3.200x"))

tempH2 <- read.table("HFGR_all_incDD2select.PASS.miss0.5.LMRG.HAP.DISCORDS.HFII.temporal.txt",stringsAsFactors = F,sep="\t")
colnames(tempH2)<-c("chr","pos","type","alleles","DD2.2D4", "HFGRII.2.10x","HFGRII.1.30x"  , "HFGRII.1.200x" , "HFGRII.2.200x" , "HFGRII.2.300x","result","freq")
genotabH2[(genotabH2$chr %in% tempH2$chr & genotabH2$pos %in% tempH2$pos),]

genotabH2 <- merge(genotabH2,subset(tempH2,result=="mut1",c("chr","pos","result")),all.x=T)

```

```{r}
cols=c("grey","dark green","orange","dark red")
names(cols)<-c(0,1,2,3)
colsc <- scale_color_manual(values = cols)
```

```{r}
#HFGR - all

#all vars
ggplot(genotab1,aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(size=3) + 
  geom_line(data=subset(genotab1,value!=0)) + facet_grid(chr ~ .) + xlim(0,3.5e6) +colsc

#SNPs only
ggplot(subset(genotab1,type=="SNP"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotab1,value!=0 & type=="SNP")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc
#INDELs only
ggplot(subset(genotab1,type=="INDEL"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotab1,value!=0 & type=="INDEL")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc

```


```{r}
#HFGR - GII

#all vars
ggplot(genotabH2,aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(size=3) + 
  geom_point(data=subset(genotabH2,result=="mut1" & value != 0),colour="red") +
  geom_line(data=subset(genotabH2,value!=0)) + facet_grid(chr ~ .) + xlim(0,3.5e6) +colsc

#SNPs only
ggplot(subset(genotabH2,type=="SNP"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH2,value!=0 & type=="SNP")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc
#INDELs only
ggplot(subset(genotabH2,type=="INDEL"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH2,value!=0 & type=="INDEL")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc

```

```{r}
#HFGR - GIII

#all vars
ggplot(genotabH3,aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(size=3) + 
  geom_line(data=subset(genotabH3,value!=0)) + facet_grid(chr ~ .) + xlim(0,3.5e6) +colsc

#SNPs only
ggplot(subset(genotabH3,type=="SNP"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH3,value!=0 & type=="SNP")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc
#INDELs only
ggplot(subset(genotabH3,type=="INDEL"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH3,value!=0 & type=="INDEL")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc

```

```{r}
#HFGR - all

#all vars
ggplot(genotabH2H3,aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(size=3) + 
  geom_line(data=subset(genotabH2H3,value!=0)) + facet_grid(chr ~ .) + xlim(0,3.5e6) +colsc

#SNPs only
ggplot(subset(genotabH2H3,type=="SNP"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH2H3,value!=0 & type=="SNP")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc
#INDELs only
ggplot(subset(genotabH2H3,type=="INDEL"),aes(x=pos,y=indv,colour=value,group=haplo,shape=type)) + geom_point(alpha=0.6,size=3) + 
  geom_line(data=subset(genotabH2H3,value!=0 & type=="INDEL")) + facet_grid(chr ~ type) + xlim(0,3.5e6) +colsc

```


